server/src/main/java/org/opensearch/index/shard/CloseableRetryableRefreshListener.java

```diff
@@ -0,0 +1,119 @@
+ * is closed, all the permits are acquired and there are no available permits to afterRefresh. This abstract class provides
+ * necessary abstract methods to schedule retry.
+ */
+public abstract class CloseableRetryableRefreshListener implements ReferenceManager.RefreshListener, Closeable {
+
+    private static final int TOTAL_PERMITS = Integer.MAX_VALUE;
+
+    private final Semaphore semaphore = new Semaphore(TOTAL_PERMITS);
+
+    private final ThreadPool retryThreadPool;
+
+    public CloseableRetryableRefreshListener(ThreadPool retryThreadPool) {
+        this.retryThreadPool = retryThreadPool;
+    }
+
+    @Override
+    public final void afterRefresh(boolean didRefresh) throws IOException {
+        if (semaphore.tryAcquire()) {
+            boolean shouldRetry;
+            try {
+                shouldRetry = performAfterRefresh(didRefresh);
+            } finally {
+                semaphore.release();
+            }
+            scheduleRetry(shouldRetry);
+        }
+    }
+
+    protected String getRetryThreadPoolName() {
+        return null;
+    }
+
+    protected TimeValue getNextRetryInterval() {
+        return null;
+    }
+
+    private void scheduleRetry(TimeValue interval, String retryThreadPool) {
+        if (this.retryThreadPool == null
+            || interval == null
+            || retryThreadPool == null
+            || ThreadPool.THREAD_POOL_TYPES.containsKey(retryThreadPool) == false
+            || interval == TimeValue.MINUS_ONE) {
+            return;
+        }
+        this.retryThreadPool.schedule(() -> {
+            if (semaphore.tryAcquire()) {
+                boolean shouldRetry;
+                try {
+                    shouldRetry = doRetry();
+                } finally {
+                    semaphore.release();
+                }
+                scheduleRetry(shouldRetry);
+            }
+        }, interval, retryThreadPool);
+    }
+
+    /**
+     * Schedules the retry based on the {@code shouldRetry} value.
+     * @param shouldRetry determines if the retry should happen.
+     */
+    private void scheduleRetry(boolean shouldRetry) {
+        if (shouldRetry) {
+            scheduleRetry(getNextRetryInterval(), getRetryThreadPoolName());
+        }
+    }
+
+    /**
+     * Override this method to provide the code that needs to be run on retry.
+     *
+     * @return true if a retry is needed. By default, it returns false which means there are no retries.
+     */
+    protected boolean doRetry() {
+        // By default it is no-op so that the listeners that do not need this, dont have to implement this.
+        return false;
+    }
+
+    /**
+     * This method needs to be overridden and be provided with what needs to be run on after refresh.
+     *
+     * @param didRefresh true if the refresh opened a new reference
+     * @return true if a retry is needed else false.
+     */
+    protected abstract boolean performAfterRefresh(boolean didRefresh);
```

Comment:
The method signature is not intuitive. 
1. `performAfterRefresh` returning boolean - here method name does not provide any details on what true/false would mean. For example `isEmpty()` returning boolean makes it clear.
2. Using `true` return value - retry would happen if something goes wrong in `performAfterRefresh`. If we assume return type indicates status of method execution, we should retry on false.

Does it make sense to have void return type and retry on exception?

Reply:
Made the changes as suggested.